src/iconforge.rs

```diff
@@ -0,0 +1,551 @@
+// 255 is invalid.
+const DIR_TO_INDEX: [u8; 11] = [255, 1, 0, 255, 2, 6, 4, 255, 3, 7, 5];
+
+
+byond_fn!(fn iconforge_generate(file_path, spritesheet_name, sprites) {
+    let file_path = file_path.to_owned();
+    let spritesheet_name = spritesheet_name.to_owned();
+    let sprites = sprites.to_owned();
+    Some(match catch_panic(&file_path, &spritesheet_name, &sprites) {
+        Ok(o) => o.to_string(),
+        Err(e) => e.to_string()
+    })
+});
+
+
+byond_fn!(fn iconforge_generate_async(file_path, spritesheet_name, sprites) {
+    // Take ownership before passing
+    let file_path = file_path.to_owned();
+    let spritesheet_name = spritesheet_name.to_owned();
+    let sprites = sprites.to_owned();
+    Some(jobs::start(move || {
+        match catch_panic(&file_path, &spritesheet_name, &sprites) {
+            Ok(o) => o.to_string(),
+            Err(e) => e.to_string()
+        }
+    }))
+});
+
+byond_fn!(fn iconforge_check(id) {
+    Some(jobs::check(id))
+});
+
+#[derive(Serialize)]
+struct Returned {
+    sizes: Vec<String>,
+    sprites: HashMap<String, SpritesheetEntry>,
+    error: String,
+}
+
+#[derive(Serialize, Clone)]
+struct SpritesheetEntry {
+    size_id: String,
+    position: u32,
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+struct IconObject {
+	icon_file: String,
+	icon_state: String,
+	dir: u8,
+	frame: u32,
+	moving: u8,
+	transform: Vec<Transform>
+}
+
+impl IconObject {
+    fn to_icostring(&self) -> Result<String, Error> {
+        zone!("to_icostring");
+        string_hash("xxh64", &serde_json::to_string(self).unwrap())
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+#[serde(tag = "type")]
+enum Transform {
+    BlendColor {
+        color: String,
+        blend_mode: u8,
+    },
+    BlendIcon {
+        icon: IconObject,
+        blend_mode: u8,
+    },
+    Scale {
+        width: u32,
+        height: u32,
+    },
+    Crop {
+        x1: i32,
+        y1: i32,
+        x2: i32,
+        y2: i32,
+    }
+}
+
+fn catch_panic(file_path: &str, spritesheet_name: &str, sprites: &str) -> std::result::Result<String, Error> {
+    let x = std::panic::catch_unwind(|| {
+        let result = generate_spritesheet(file_path, spritesheet_name, sprites);
+        frame!();
+        result
+    });
+    if let Err(err) = x {
+        let message: Option<String> = err
+            .downcast_ref::<&'static str>()
+            .map(|payload| payload.to_string())
+            .or_else(|| {
+                err.downcast_ref::<String>().cloned()
+            });
+        return Err(Error::IconState(message.unwrap().to_owned()));
+    }
+    x.ok().unwrap()
+}
+
+fn generate_spritesheet(file_path: &str, spritesheet_name: &str, sprites: &str) -> std::result::Result<String, Error> {
+    zone!("generate_spritesheet");
+
+    let error: Arc<Mutex<Vec<String>>> = Arc::new(Mutex::new(Vec::new()));
```

Comment:
For a lot of this stuff, you don't need the explicit types written out when initialized like this